### DISCUSSION TITLE
Paste >=19.7.0 compatibility with Twilio CLI

### DISCUSSION THREAD

**Discussion by dmitriy-kudelko** [twilio-labs/paste#3416] [Q&A] [answered]
Hey guys 👋 

We're using Paste to develop custom plugins for Flex. We noticed that upgrading from `19.6.0` to `19.7.0` and onwards brakes the build throwing e.g.:
```
Module parse failed: Unexpected token (1:1537)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
```

We're using `@twilio/flex-plugin-scripts@6.1.2` which is a latest stable version at the moment.

Is it a known issue?  Do we need to customize the default Webpack configuration in some way to avoid that?

Thanks!

**Comment by SiTaggart** [comment]
Hi @dmitriy-kudelko 

To double clarify, that's seen in 19.7.0, not just v20? v20 I might expect something like that as we changed the compilation target to browser, not node, but 19.7 is a little out of the blue.

Is there also a longer error message or log you can share. That currently one doesn't tell us too much to work with. 

Thanks

**Reply by dmitriy-kudelko** [reply]
@SiTaggart , yes, I've checked, and this affects both v19.7.0 and v20.

The easiest way to reproduce it would be to install Twilio CLI, Twilio Flex Plugins CLI. Then generate a new Typescript Flex plugin project and import some components from Paste:

https://www.twilio.com/docs/flex/quickstart/getting-started-plugin#set-up-a-flex-environment-with-flex-plugins-cli

As for the error, here's the full one:

```js
./node_modules/@twilio-paste/radio-group/dist/index.es.js 1:1537
Module parse failed: Unexpected token (1:1537)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> import*as React2 from"react";import"prop-types";import{useUID}from"@twilio-paste/uid-library";import{Box}from"@twilio-paste/box";import{BaseRadioCheckboxControl,BaseRadioCheckboxLabel,BaseRadioCheckboxLabelText,BaseRadioCheckboxHelpText}from"@twilio-paste/base-radio-checkbox";import*as React from"react";var RadioContext=React.createContext({name:"",value:"",disabled:!1,hasError:!1,onChange:()=>{}});var HiddenRadio=React2.forwardRef((props,ref)=>React2.createElement(Box,{as:"input",type:"radio",size:"size0",border:"none",overflow:"hidden",padding:"space0",margin:"space0",whiteSpace:"nowrap",textTransform:"none",position:"absolute",clip:"rect(0 0 0 0)",ref,.props}));HiddenRadio.displayName="HiddenRadio";var Radio=React2.forwardRef(({id,name,element="RADIO",value,checked,defaultChecked,disabled,hasError,onChange,children,helpText,.props},ref)=>{if(checked!=null&&defaultChecked!=null)throw new Error("[Paste Radio] Do not provide both 'defaultChecked' and 'checked' to Radio at the same time. Please consider if you want this component to be controlled or uncontrolled.");let[checkedState,setCheckedState]=React2.useState(defaultChecked),radioGroupContext=React2.useContext(RadioContext),helpTextId=useUID(),radioId=id||useUID(),isControlled=React2.useMemo(()=>checked!==void 0||radioGroupContext.value!=="",[]),handleChange=React2.useCallback(event=>{onChange?onChange(event):radioGroupContext.onChange(event),isControlled||setCheckedState(event.target.checked)},[onChange,radioGroupContext,isControlled]),state={name:name??radioGroupContext.name,disabled:disabled??radioGroupContext.disabled,hasError:hasError??radioGroupContext.hasError};isControlled?radioGroupContext.value!==""?state.checked=radioGroupContext.value===value:state.checked=checked:state.defaultChecked=defaultChecked;let mergedChecked=isControlled?state.checked||state.defaultChecked:checkedState;return React2.createElement(Box,{element,position:"relative",display:"inline-flex",alignItems:"flex-start",flexDirection:"column",verticalAlign:"top"},React2.createElement(HiddenRadio,{.props,.state,value,"aria-describedby":helpTextId,"aria-invalid":state.hasError,onChange:handleChange,id:radioId,ref}),React2.createElement(BaseRadioCheckboxLabel,{disabled:state.disabled,htmlFor:radioId},React2.createElement(BaseRadioCheckboxControl,{element:`${element}CONTROL`,borderRadius:"borderRadiusCircle",disabled:state.disabled,type:"radio",checkedAndDisabledSibling:{color:"colorTextWeaker"}},React2.createElement(Box,{as:"span",element:`${element}CONTROLCIRCLE`,lineHeight:"lineHeight0",display:mergedChecked?"block":"none",color:"inherit",size:"sizeIcon10"},React2.createElement("svg",{role:"img","aria-hidden":!0,width:"100%",height:"100%",viewBox:"0 0 16 16",fill:"none"},React2.createElement("circle",{fill:"currentColor",cx:8,cy:8,r:3})))),React2.createElement(BaseRadioCheckboxLabelText,{element:`${element}LABELTEXT`,fontWeight:"fontWeightMedium"},children)),helpText&&React2.createElement(BaseRadioCheckboxHelpText,{element:`${element}HELPTEXTWRAPPER`,helpTextId},helpText))});Radio.displayName="Radio";import*as React3 from"react";import PropTypes2 from"prop-types";import{InlineControlGroup}from"@twilio-paste/inline-control-group";var RadioGroup=React3.forwardRef(({name,element="RADIOGROUP",value,onChange,disabled=!1,errorText,children,i18nRequiredLabel="(required)",.props},ref)=>{let onChangeHandler=React3.useCallback(event=>{onChange?.(event.target.value)},[onChange]),contextValue=React3.useMemo(()=>({name,value:value||"",disabled,hasError:errorText!=null,onChange:onChangeHandler}),[name,value,disabled,errorText,onChangeHandler]);return React3.createElement(RadioContext.Provider,{value:contextValue},React3.createElement(InlineControlGroup,{element,.props,disabled,errorText,name,ref,i18nRequiredLabel},children))});RadioGroup.displayName="RadioGroup";RadioGroup.propTypes={name:PropTypes2.string.isRequired,onChange:PropTypes2.func,value:PropTypes2.string,element:PropTypes2.string,i18nRequiredLabel:PropTypes2.string};export{HiddenRadio,Radio,RadioGroup};
|

```

**Reply by SiTaggart** [reply]
Did you also happen to update `@twilio/flex-plugin-scripts`? Or was it just the Paste packages?

**Reply by dmitriy-kudelko** [reply]
@SiTaggart, no the scripts package wasn't updated because

> We're using @twilio/flex-plugin-scripts@6.1.2 which is a latest stable version at the moment.

**Reply by SiTaggart** [reply]
Can you try the latest alpha build, looks like the compilation target in that might be more modern https://www.npmjs.com/package/@twilio/flex-plugin-scripts/v/6.1.3-alpha.202308111018

**Reply by dmitriy-kudelko** [reply]
Nope, this doesn't seem to help.

**Reply by SiTaggart** [reply]
OK, we'll continue to take a look. This will likely require a Webpack or Flex Plugin update. Essentially we've just moved to a slightly more modern flavor of JS which removed some shims for older browsers which aren't actually supported anymore by Flex. Just the tools haven't elvolved with us.

If 19.6 works for now, you're not missing much in terms of features, and you can get building whilst we work with the flex team

**Reply by TheSisb** [reply]
Hey @dmitriy-kudelko,

I took a look at the differences in the file from `radio-group/dist/index.es.js` before version 19.7 and after.  Nothing stands out other than removing the spread operator shims.

[Before file](https://gist.github.com/TheSisb/86783a942df5c23123ec821252484b1a)

[After file](https://gist.github.com/TheSisb/93f35ef2f3d116353f89a51275e03406)

You can compare them in [an online diff tool](https://www.diffchecker.com/).

As far as I can tell, there shouldn't be any significant changes in the file.  It seems like some likely options are:

1) your webpack config is suddenly loading our `*.es.js` dist file instead of the commonjs `*.js` dist file
2) your webpack config suddenly stopped supporting `*.es.js` files

If possible, get to a working state in your code and _only_ update the Paste version without making any other changes. Does the issue still occur?

**Reply by dmitriy-kudelko** [reply]
Hi, @TheSisb,

Well, it was only change that we did. Nothing else was updated. It was actually a renovate bot PR.

Our webpack config customization is minimal, so it's almost a default one from flex plugins scripts. 

```js
const path = require('path');
const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer');

module.exports = (config) => {
  /**
   * Customize the webpack by modifying the config object.
   * Consult https://webpack.js.org/configuration for more information
   */

  config.module.rules.unshift({
    test: /\.inline.svg$/,
    use: ['@svgr/webpack'],
  });

  if (process.env.ANALYZE_BUNDLE) {
    config.plugins.push(new BundleAnalyzerPlugin());
  }

  Object.assign(config.resolve.alias, {
    '@': path.resolve(__dirname, 'src'),
    '~': path.resolve(__dirname, '.'),
  });

  return config;
};
```

**Reply by dmitriy-kudelko** [reply]
@SiTaggart, yes, absolutely. We can for now just skip the updates. It's not a blocker for us. Just wanted to let you guys know, 'cause my impression was that the compatibility with Flex plugins CLI is implied as Paste is a recommended UI library for building Flex plugins.

So this would likely more affect the ones who start building a new plugin, follow instructions for installing Twilio Flex plugins CLI, install Paste  and realize that the setup doesn't work.

**Reply by SiTaggart** [reply, marked as answer]
Hi @dmitriy-kudelko 

So, in `v19.7.0` we [bumped our browser targets](https://github.com/twilio-labs/paste/commit/c4a70d95bf8d61649e5fb3cb9b67148438d27662#diff-9f9e885226db414623d1da691239d618c81fbec07292a2143ef22c59eb27eca5R60) from `target: ['chrome66', 'firefox58', 'safari11', 'edge79', 'node12.19.0'],` to `target: ['chrome100', 'firefox100', 'safari14', 'edge100', 'node18.16.0'],`

This helps make our bundles a little smaller as modern browsers and node now support many more JavaScript features that would no longer require being transpiled when we compile our typescript.

What this has resulted in is the removal of the object spread and object rest spread shims from our compiled JS. That, by the looks of it, is what is causing the Flex plugin builder to have problems. You might need to update that webpack config to transpile our code before it gets run, or have Flex update the plugin builder to support more modern JavaScript, so you don't have to.

This seems logical to us as Flex is only supported in Chrome which has great spread support, and Node 16 is about to go [end of Life in September](https://nodejs.org/en/blog/announcements/nodejs16-eol).

If you need any further support, we'd probably direct you to the Flex product team from there on, as they'd be in a much better position in supporting Flex for you.

**Reply by SiTaggart** [reply]
> So this would likely more affect the ones who start building a new plugin, follow instructions for installing Twilio Flex plugins CLI, install Paste and realize that the setup doesn't work.

Agreed. You also need to do quite a bit of extra work from the quick start guide to even get towards using a newer version of paste, as the Flex-UI package pins a couple of Paste deps to Paste Core v15. You have to wrestle some resolutions to get it picking up the right ones.

Hopefully they will improve that experience in the coming months

**Reply by dmitriy-kudelko** [reply]
@SiTaggart, alright, thank you!

It would be enough for now to just let the Flex team know about this issue.

**Reply by moretalk** [reply]
Hi @SiTaggart @dmitriy-kudelko wanted to add to this. So our customers could install any combination of upto 100 plugins in there Flex Environment. I am noticing that our Flex 1.X plugins are at most 1mb (because we do not need to bundle Material UI) whereas our Flex 2.X plugins are close to double this because we need to bundle newer versions of Paste in each plugin.

It would be far better if the Flex Team allowed for a choice of Paste Version (in the same way as they do React). This way we could support newer versions of Paste.

The Docs are in desperate need of updating. I did find that I could not get any higher than 17.7.0. Version 18.X said that it could not be used as JSX? and version 19 refused to bundle in webpack.

**Reply by serifluous** [reply]
Hey @moretalk ! For this issue and the ones you mentioned in #2752 , can you create a support ticket with the Flex team, or speak with your account manager if you have one? They'll be the best people who can help you out with this.
